### ISSUE TITLE
swaybar renders fcitx's tray icon as red sad face

### ISSUE THREAD

**Issue by Arnavion** [swaywm/sway#5645]
(Copying from <https://github.com/swaywm/sway/issues/4825#issuecomment-678673811> on emersion's request to not reuse that issue.)

- **Sway Version:** `sway version v1.5`

- **Debug Log:** <https://gist.github.com/Arnavion/0718ea34304e9188d5d5197ef651585c#file-sway-debug-log> (`swaybar` was also run with `-d`)

- **Configuration File:** Minimal config <https://gist.github.com/Arnavion/0718ea34304e9188d5d5197ef651585c#file-sway-debug-config>

- **Stack Trace:** None

- **Description:**

swaybar renders fcitx's tray icon as a red sad face just like in <https://github.com/swaywm/sway/issues/4825>:

```sh
$ grep -i tray sway-debug.log

00:00:00.115 [swaybar/tray/tray.c:37] Initializing tray
00:00:00.116 [swaybar/tray/watcher.c:202] Registered org.freedesktop.StatusNotifierWatcher
00:00:00.116 [swaybar/tray/watcher.c:186] Failed to acquire service name 'org.kde.StatusNotifierWatcher':another tray is already running
00:00:00.116 [swaybar/tray/host.c:198] Registered org.freedesktop.StatusNotifierHost-93313
00:00:00.117 [swaybar/tray/host.c:198] Registered org.kde.StatusNotifierHost-93313
00:00:00.117 [swaybar/tray/icon.c:313] Failed to load theme 'default' - parsing of file '/home/arnavion/.icons/default/index.theme' failed on line 5 (group 'Icon Theme'): missing required key 'Directories'
00:00:00.120 [swaybar/tray/icon.c:379] Loaded icon themes: Hicolor, Adwaita
00:00:00.132 00:00:00.435 [swaybar/tray/watcher.c:107] Registering Status Notifier Host 'org.freedesktop.StatusNotifierHost-93313'
00:00:00.132 [swaybar/tray/host.c:24] Registering Status Notifier Item 'org.kde.StatusNotifierItem-99416-1/StatusNotifierItem'
00:00:00.132 [swaybar/tray/item.c:152] org.kde.StatusNotifierItem-99416-1/StatusNotifierItem IconThemePath = ''
00:00:00.132 [swaybar/tray/item.c:152] org.kde.StatusNotifierItem-99416-1/StatusNotifierItem Status = 'Active'
00:00:00.132 [swaybar/tray/item.c:152] org.kde.StatusNotifierItem-99416-1/StatusNotifierItem IconName = 'input-keyboard'
00:00:00.132 [swaybar/tray/item.c:121] org.kde.StatusNotifierItem-99416-1/StatusNotifierItem IconPixmap: No such property ('org.kde.StatusNotifierItem.IconPixmap')
00:00:00.132 [swaybar/tray/item.c:152] org.kde.StatusNotifierItem-99416-1/StatusNotifierItem AttentionIconName = ''
00:00:00.132 [swaybar/tray/item.c:121] org.kde.StatusNotifierItem-99416-1/StatusNotifierItem AttentionIconPixmap: No such property ('org.kde.StatusNotifierItem.AttentionIconPixmap')
00:00:00.132 [swaybar/tray/item.c:121] org.kde.StatusNotifierItem-99416-1/StatusNotifierItem ItemIsMenu: No such property ('org.kde.StatusNotifierItem.ItemIsMenu')
00:00:00.132 [swaybar/tray/item.c:152] org.kde.StatusNotifierItem-99416-1/StatusNotifierItem Menu = '/MenuBar'
00:00:02.062 [swaybar/tray/host.c:24] Registering Status Notifier Item 'org.kde.StatusNotifierItem-93317-1/StatusNotifierItem'
00:00:02.063 [swaybar/tray/item.c:152] org.kde.StatusNotifierItem-93317-1/StatusNotifierItem IconThemePath = ''
00:00:02.063 [swaybar/tray/item.c:152] org.kde.StatusNotifierItem-93317-1/StatusNotifierItem Status = 'Active'
00:00:02.063 [swaybar/tray/item.c:152] org.kde.StatusNotifierItem-93317-1/StatusNotifierItem IconName = 'input-keyboard'
00:00:02.072 [swaybar/tray/item.c:121] org.kde.StatusNotifierItem-93317-1/StatusNotifierItem IconPixmap: No such property ('org.kde.StatusNotifierItem.IconPixmap')
00:00:02.072 [swaybar/tray/item.c:152] org.kde.StatusNotifierItem-93317-1/StatusNotifierItem AttentionIconName = ''
00:00:02.072 [swaybar/tray/item.c:121] org.kde.StatusNotifierItem-93317-1/StatusNotifierItem AttentionIconPixmap: No such property ('org.kde.StatusNotifierItem.AttentionIconPixmap')
00:00:02.072 [swaybar/tray/item.c:121] org.kde.StatusNotifierItem-93317-1/StatusNotifierItem ItemIsMenu: No such property ('org.kde.StatusNotifierItem.ItemIsMenu')
00:00:02.072 [swaybar/tray/item.c:152] org.kde.StatusNotifierItem-93317-1/StatusNotifierItem Menu = '/MenuBar'
```

(There are two icons because the first one is from fcitx running in my original tty with a non-debug sway session, and the second is from the fcitx `exec_always`'d into the debug sway session.)

fcitx according to dbus has:

![image](https://user-images.githubusercontent.com/1096010/90962729-01bab880-e467-11ea-9948-3e79340db960.png)

(Note: This image is from the original issue, which is why the dbus names doesn't match.)

To clarify for anyone not familiar with d-feet's UI: `.IconName` is set to `"input-keyboard"`, `.IconThemePath` is set to `""`. I did not query the other properties which is why they don't show their values.

Looking at `swaybar/tray/item.c`, the error logs are benign since it just gets whatever properties it can. It does look up icon using the value it got from `IconName`, so I'm not sure why it's failing to find it.

```sh
$ find /usr/share/icons -name 'input-keyboard.*'

/usr/share/icons/Adwaita/16x16/legacy/input-keyboard.png
/usr/share/icons/Adwaita/22x22/legacy/input-keyboard.png
/usr/share/icons/Adwaita/24x24/legacy/input-keyboard.png
/usr/share/icons/Adwaita/256x256/legacy/input-keyboard.png
/usr/share/icons/Adwaita/32x32/legacy/input-keyboard.png
/usr/share/icons/Adwaita/48x48/legacy/input-keyboard.png
```

FWIW the tray icon renders fine with waybar. It appears to be using the `/usr/share/icons/Adwaita/16x16/legacy/input-keyboard.png` one.

And in case that index.theme error is relevant:

```sh
$ cat ~/.icons/default/index.theme

# This file is written by LXAppearance. Do not edit.
[Icon Theme]
Name=Default
Comment=Default Cursor Theme
Inherits=Adwaita
```

**Comment by ianyfan** [comment]
Set your bar's icon theme to Adwaita

**Comment by Arnavion** [comment]
Thanks. That works, and also makes sense:

>```
>icon_theme <name>
>
>    Sets the icon theme that sway will look for item icons in. This option has no default value, because sway will always default to the fallback theme, hicolor.
>```

I guess waybar works by default because it gets the icon theme from the general GTK configuration.

**Comment by ianyfan** [comment]
It would probably help to add a debug message about which icon theme is being used.

**Comment by amano-kenji** [comment]
This issue appeared again after the latest system update. The icon theme is Adwaita.

**Comment by TinusgragLin** [comment]
> This issue appeared again after the latest system update. The icon theme is Adwaita.

Happened to me this morning after I updated my Arch. I tried installing a new random icon theme (`cutefish-icons`, specifically) and using the new icon theme instead. Turns out it works!

**Comment by neilbags** [comment]
I fixed this in Ubuntu 23.10 by installing ```adwaita-icon-theme-full``` and setting ```gtk-icon-theme-name=Adwaita``` in ```.config/gtk-3.0/settings.ini```
